File: internal/tools/sessions.go
package tools

import (
	"context"
	"encoding/json"
	"fmt"
)

func (r *Registry) registerSessionTools() {
	r.register(ToolDef{
		Name:        "session_create",
		Description: "Launch a new stealth browser session with anti-detection, proxy, and fingerprinting",
		InputSchema: map[string]any{
			"type": "object",
			"properties": map[string]any{
				"proxy_type":    map[string]any{"type": "string", "enum": []string{"residential", "datacenter", "isp", "mobile"}, "description": "Proxy type to use"},
				"proxy_country": map[string]any{"type": "string", "description": "ISO country code for proxy geo-targeting (e.g., US, GB, DE)"},
				"stealth":       map[string]any{"type": "string", "enum": []string{"none", "standard", "max"}, "default": "max", "description": "Anti-detection level"},
				"profile_id":    map[string]any{"type": "string", "description": "Load a saved browser profile (cookies, storage, fingerprint)"},
				"viewport":      map[string]any{"type": "object", "properties": map[string]any{"width": map[string]any{"type": "integer"}, "height": map[string]any{"type": "integer"}}},
			},
		},
		Handler: r.handleSessionCreate,
	})

	r.register(ToolDef{
		Name:        "session_list",
		Description: "List all active browser sessions",
		InputSchema: map[string]any{
			"type":       "object",
			"properties": map[string]any{},
		},
		Handler: r.handleSessionList,
	})

	r.register(ToolDef{
		Name:        "session_get",
		Description: "Get details about a specific session including CDP endpoint",
		InputSchema: map[string]any{
			"type": "object",
			"properties": map[string]any{
				"session_id": map[string]any{"type": "string", "description": "Session ID"},
			},
			"required": []string{"session_id"},
		},
		Handler: r.handleSessionGet,
	})

	r.register(ToolDef{
		Name:        "session_destroy",
		Description: "Destroy a browser session and clean up resources",
		InputSchema: map[string]any{
			"type": "object",
			"properties": map[string]any{
				"session_id":   map[string]any{"type": "string", "description": "Session ID"},
				"save_profile": map[string]any{"type": "boolean", "description": "Save session state (cookies, storage) to profile before destroying"},
			},
			"required": []string{"session_id"},
		},
		Handler: r.handleSessionDestroy,
	})
}

func (r *Registry) handleSessionCreate(ctx context.Context, args map[string]any) (*ToolResult, error) {
	body := map[string]any{
		"stealth": "max",
	}
	if v, ok := args["proxy_type"]; ok {
		proxy := map[string]any{"type": v}
		if country, ok := args["proxy_country"]; ok {
			proxy["country"] = country
		}
		body["proxy"] = proxy
	}
	if v, ok := args["stealth"]; ok {
		body["stealth"] = v
	}
	if v, ok := args["profile_id"]; ok {
		body["profile_id"] = v
	}
	if v, ok := args["viewport"]; ok {
		body["viewport"] = v
	}

	resp, err := r.client.Do(ctx, "POST", "/v1/sessions", body)
	if err != nil {
		return nil, fmt.Errorf("creating session: %w", err)
	}

	var result map[string]any
	if err := json.Unmarshal(resp, &result); err != nil {
		return nil, fmt.Errorf("parsing response: %w", err)
	}

	return jsonResult(result)
}

func (r *Registry) handleSessionList(ctx context.Context, _ map[string]any) (*ToolResult, error) {
	resp, err := r.client.Do(ctx, "GET", "/v1/sessions", nil)
	if err != nil {
		return nil, fmt.Errorf("listing sessions: %w", err)
	}

	var result any
	if err := json.Unmarshal(resp, &result); err != nil {
		return nil, fmt.Errorf("parsing response: %w", err)
	}

	return jsonResult(result)
}

func (r *Registry) handleSessionGet(ctx context.Context, args map[string]any) (*ToolResult, error) {
	sessionID, _ := args["session_id"].(string)
	if sessionID == "" {
		return nil, fmt.Errorf("session_id is required")
	}

	resp, err := r.client.Do(ctx, "GET", "/v1/sessions/"+sessionID, nil)
	if err != nil {
		return nil, fmt.Errorf("getting session: %w", err)
	}

	var result map[string]any
	if err := json.Unmarshal(resp, &result); err != nil {
		return nil, fmt.Errorf("parsing response: %w", err)
	}

	return jsonResult(result)
}

func (r *Registry) handleSessionDestroy(ctx context.Context, args map[string]any) (*ToolResult, error) {
	sessionID, _ := args["session_id"].(string)
	if sessionID == "" {
		return nil, fmt.Errorf("session_id is required")
	}

	var body map[string]any
	if save, ok := args["save_profile"].(bool); ok && save {
		body = map[string]any{"save_profile": true}
	}

	resp, err := r.client.Do(ctx, "DELETE", "/v1/sessions/"+sessionID, body)
	if err != nil {
		return nil, fmt.Errorf("destroying session: %w", err)
	}

	var result map[string]any
	if err := json.Unmarshal(resp, &result); err != nil {
		return nil, fmt.Errorf("parsing response: %w", err)
	}

	return jsonResult(result)
}

File: src/tools/sessions.ts
import type { Tool } from "@modelcontextprotocol/sdk/types.js";
import type { MeshbrowClient } from "../client.js";

export const sessionTools: Tool[] = [
  {
    name: "session_create",
    description:
      "Create a new stealth browser session. Returns session ID and CDP endpoint for connecting with Playwright/Puppeteer.",
    inputSchema: {
      type: "object" as const,
      properties: {
        stealth: {
          type: "string",
          enum: ["none", "basic", "max"],
          description: "Stealth level for anti-detection (default: max)",
        },
        proxy_type: {
          type: "string",
          enum: ["residential", "datacenter", "isp", "mobile"],
          description: "Proxy type to use (default: residential)",
        },
        proxy_country: {
          type: "string",
          description: "Proxy country code, ISO 3166-1 alpha-2 (e.g., US, GB, DE)",
        },
        viewport_width: {
          type: "number",
          description: "Browser viewport width in pixels (default: 1920)",
        },
        viewport_height: {
          type: "number",
          description: "Browser viewport height in pixels (default: 1080)",
        },
        locale: {
          type: "string",
          description: "Browser locale (e.g., en-US)",
        },
        timezone: {
          type: "string",
          description: "Browser timezone (e.g., America/New_York)",
        },
        timeout: {
          type: "number",
          description: "Session timeout in minutes (default: 30)",
        },
        profile_id: {
          type: "string",
          description: "Profile ID to restore session state from",
        },
      },
    },
  },
  {
    name: "session_list",
    description: "List all active browser sessions with their status and configuration.",
    inputSchema: {
      type: "object" as const,
      properties: {},
    },
  },
  {
    name: "session_get",
    description: "Get details of a specific browser session including CDP endpoint.",
    inputSchema: {
      type: "object" as const,
      properties: {
        session_id: {
          type: "string",
          description: "The session ID to get details for",
        },
      },
      required: ["session_id"],
    },
  },
  {
    name: "session_destroy",
    description: "Destroy a browser session. Optionally save the session profile for reuse.",
    inputSchema: {
      type: "object" as const,
      properties: {
        session_id: {
          type: "string",
          description: "The session ID to destroy",
        },
        save_profile: {
          type: "boolean",
          description: "Whether to save cookies/state to the profile before destroying",
        },
      },
      required: ["session_id"],
    },
  },
];

export async function handleSessionTool(
  client: MeshbrowClient,
  name: string,
  params: Record<string, unknown>
): Promise<unknown> {
  switch (name) {
    case "session_create": {
      const body: Record<string, unknown> = {
        stealth: params.stealth || "max",
        timeout: params.timeout || 30,
      };

      if (params.proxy_type || params.proxy_country) {
        body.proxy = {
          type: params.proxy_type || "residential",
          country: params.proxy_country,
        };
      }

      if (params.viewport_width || params.viewport_height) {
        body.viewport = {
          width: params.viewport_width || 1920,
          height: params.viewport_height || 1080,
        };
      }

      if (params.locale) body.locale = params.locale;
      if (params.timezone) body.timezone = params.timezone;
      if (params.profile_id) body.profile_id = params.profile_id;

      return client.post("/v1/sessions", body);
    }

    case "session_list":
      return client.get("/v1/sessions");

    case "session_get":
      return client.get(`/v1/sessions/${params.session_id}`);

    case "session_destroy": {
      const query = params.save_profile ? "?save_profile=true" : "";
      return client.delete(`/v1/sessions/${params.session_id}${query}`);
    }

    default:
      throw new Error(`Unknown session tool: ${name}`);
  }
}
